### ISSUE TITLE
message format issue

### ISSUE THREAD

**Issue by newhook** [segmentio/kafka-go#100]
I'm grabbing messages off our kafka stream and noticed an issue with the message content. After decompressing the message (snappy encoded) the data appears to have some leading junk. In hex:

`0 0 0 0 2 c3 56 83 0 0 5 23 67 c 0 4 0 0 ff ff ff ff 0 0 5 15`

Any ideas what this is? I thought it might be a message batch as documented here:

https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-CommonRequestandResponseStructure

But there are not enough bytes for that.

**Comment by achille-roussel** [comment]
Do you mind giving a bit more info about the version of Kafka you're running? Also if you have some code to reproduce the issue in isolation that would be a good start for me to dig into the issue.

**Comment by newhook** [comment]
We're running kafka 2.11-1. The code I was testing with is as follows. Note that the data is snappy encoded. We originally inserted the data into kafka using kafkacat.

`kafkacat -P -b kafka-1 -t server-invalid -z snappy`

```
package main

import (
	"bytes"
	"context"
	"encoding/binary"
	"fmt"

	"github.com/golang/snappy"
	kafka "github.com/kafka-go"
)

var xerialHeader = []byte{130, 83, 78, 65, 80, 80, 89, 0}

// Decode decodes snappy data whether it is traditional unframed
// or includes the xerial framing format.
func Decode(src []byte) ([]byte, error) {
	if !bytes.Equal(src[:8], xerialHeader) {
		return snappy.Decode(nil, src)
	}

	var (
		pos   = uint32(16)
		max   = uint32(len(src))
		dst   = make([]byte, 0, len(src))
		chunk []byte
		err   error
	)
	for pos < max {
		size := binary.BigEndian.Uint32(src[pos : pos+4])
		pos += 4

		chunk, err = snappy.Decode(chunk, src[pos:pos+size])
		if err != nil {
			return nil, err
		}
		pos += size
		dst = append(dst, chunk...)
	}
	return dst, nil
}

func main() {
	r := kafka.NewReader(kafka.ReaderConfig{
		Brokers:   []string{"kafka-1:9092"},
		Topic:     "server-invalid",
		Partition: 0,
		MinBytes:  10e3, // 10KB
		MaxBytes:  10e6, // 10MB
	})
	r.SetOffset(46356099)

	for {
		m, err := r.ReadMessage(context.Background())
		if err != nil {
			fmt.Println(err)
			break
		}
		data, err := Decode(m.Value)
		if err != nil {
			fmt.Println(err)
			break
		}
		for i, c := range data {
			fmt.Printf("%x ", c)
		}
		fmt.Printf("\n")
		fmt.Printf("message at offset %d: %s = %s\n", m.Offset, string(m.Key), string(data))
		break
	}

	r.Close()
}
```

**Comment by Pryz** [comment]
Hey @newhook , are you still experiencing this issue ?

2.11-1 is the version of Scala used by Kafka but not the version of Kafka itself. The version of Kafka will be something like : `0.10.2.2`, `0.11.0.1`, `1.1.0` etc etc.

Note that `kafka-go` will handle compression once https://github.com/segmentio/kafka-go/pull/106 is merged.

**Comment by Pryz** [comment]
Hey @joepurdy, I'm hitting the same issue.

The problem is that Kafka uses the idea of `MessageSet` to be able to compress and uncompress payload more efficiently. A set is basically a message which wraps other messages so that Kafka can compress an entire batch of messages.

Note that `RecordBatch` is replacing `MessageSet` since Kafka 0.11 but the idea is the same.

We haven't implemented any of those in kafka-go yet and this is why you are having those leading characters (which are basically some message attributes).

**Comment by stevevls** [comment]
This should be fixed by #135.  Please re-open if you think it's still an issue.
